### PR TITLE
[OpenAI] Service version `v2023-12-01-preview` support

### PR DESCRIPTION
**Pull request by jpalvarezl**
## Description

- This PR is currently generating from `HEAD` of the the [PR on the PR  for TSP for v2023_12_01_preview](https://github.com/Azure/azure-rest-api-specs/pull/26790)
- I have adjusted samples based on the implementation for .NET in [this PR](https://github.com/Azure/azure-sdk-for-net/blob/9d3134b10eca189c141ed77d9b80459232fd8852/sdk/openai/Azure.AI.OpenAI/README.md)
- Do we still need `prompt_annotations` ? (That was a fix for the retroactive rename of the `prompt_filter_result`)

## Remaining TODOs
- [x] Tests need to be re-recorded
- [x] Removed legacy `ChatMessage`, `ImageResponse`, `ImageLocation` classes as well as everything `LRO` related.
- [x] Adjusted usage of `AzureCognitiveSearchChatExtensionConfiguration` for unified class infra
- [x] `ChatRequestMessageUser` ctx takes `BinaryData` as parameter:
  - [x] Accept single string ctx (now it takes `BinaryData`.
  - [x] Accept array of `ChatMessageContentItem`
- [x] match test coverage (cases, not branch) of .NET
- [x] Write samples for tool call and vision
- [x] Update README and release notes:
  - [x]  Link to `usage` samples
  - [x] Update sample titles
  - [x] Add new sections for new features

### CODE REVIEW

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/implementation/OpenAIClientImpl.java [hunk 1]
```diff
@@ -1611,57 +1894,51 @@ public Response<BinaryData> getChatCompletionsWithAzureExtensionsWithResponse(St
...
      * @throws ResourceNotFoundException thrown if the request is rejected by server on status code 404.
      * @throws ResourceModifiedException thrown if the request is rejected by server on status code 409.
-     * @return a polling status update or final response payload for an image operation along with {@link Response} on
-     * successful completion of {@link Mono}.
+     * @return the result of a successful image generation operation along with {@link Response} on successful
+     * completion of {@link Mono}.
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
-    private Mono<Response<BinaryData>> beginAzureBatchImageGenerationWithResponseAsync(
+    public Mono<Response<BinaryData>> getImageGenerationsWithResponseAsync(String deploymentOrModelName,
```
Comment (mssfang) [inline sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/implementation/OpenAIClientImpl.java:1920]: For naming consistency, we don't need to add suffix `Async` to the method name. 
Suggestion: rename `getImageGenerationsWithResponseAsync` to `getImageGenerationsWithResponse`.

Also, Just a thought, we can shorten the name to `getImagesWithResponse` if that aligns with other languages. But I saw codegen uses `xxxImageGenerations` https://github.com/Azure/azure-sdk-for-java/pull/37739/files#diff-9f3b8c373fe094bd63c683cf0026bec4e39b75490a1a0f83d1eb43dc03fac251R1514

Maybe a naming discussion could align with the TypeSpec definition in the future.

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/OpenAIClient.java [hunk 2]
```diff
@@ -1471,25 +1443,79 @@ ChatCompletions getChatCompletionsWithAzureExtensions(String deploymentOrModelNa
...
+     * @param requestOptions The options to configure the HTTP request before HTTP client sends it.
+     * @throws HttpResponseException thrown if the request is rejected by server.
+     * @throws ClientAuthenticationException thrown if the request is rejected by server on status code 401.
+     * @throws ResourceNotFoundException thrown if the request is rejected by server on status code 404.
+     * @throws ResourceModifiedException thrown if the request is rejected by server on status code 409.
+     * @return the result of a successful image generation operation along with {@link Response}.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Response<BinaryData> getImageGenerationsWithResponse(String deploymentOrModelName,
+        BinaryData imageGenerationOptions, RequestOptions requestOptions) {
```
Comment (mssfang) [inline sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/OpenAIClient.java:1490]: Feel that `BinaryData` is overused. We should have a method 
`public Response<ImageGenerations> getImageGenerationsWithResponse(String deploymentOrModelName,
        imageGenerationOptions imageGenerationOptions, RequestOptions requestOptions)`

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/OpenAIClient.java [hunk 3]
```diff
@@ -715,89 +712,6 @@ public IterableStream<ChatCompletions> getChatCompletionsStream(String deploymen
...
-     * @throws IllegalArgumentException thrown if parameters fail the validation.
-     * @throws HttpResponseException thrown if the request is rejected by server.
-     * @throws ClientAuthenticationException thrown if the request is rejected by server on status code 401.
-     * @throws ResourceNotFoundException thrown if the request is rejected by server on status code 404.
-     * @throws ResourceModifiedException thrown if the request is rejected by server on status code 409.
-     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
-     * @return the {@link ImageResponse} for the image generation result.
-     */
-    @ServiceMethod(returns = ReturnType.SINGLE)
-    public ImageResponse getImages(ImageGenerationOptions imageGenerationOptions) {
```
Comment (mssfang) [inline sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/OpenAIClient.java:731]: We can keep this convenience API for the customer who doesn't want to return with `response`.

Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/OpenAIClient.java:731]: Naming can be aligned by `getImageGenerations`, or  using `getImagesWithResponse` which drops `Generation` in the response method.

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/OpenAIClient.java (hunk 2 above)
Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/OpenAIClient.java:1490]: https://github.com/Azure/azure-sdk-for-java/pull/36859/files#diff-b477cd2968cae4b347314275298a5f5c2f49926f547869eba6060b0874adb43dR10

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/ChatChoice.java [hunk 4]
```diff
@@ -42,29 +42,15 @@ public final class ChatChoice {
...
+    private ChatResponseMessage delta;
 
     /**
      * Get the message property: The chat message for a given chat completions prompt.
      *
      * @return the message value.
      */
     @Generated
-    public ChatMessage getMessage() {
+    public ChatResponseMessage getMessage() {
```
Comment (mssfang) [inline sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/ChatChoice.java:53]: Wondering do you know why `ChatResponseMessage` is a better name than the original name `ChatMessage`? And why not using `ChatResponseChoice` to replace `ChatChoice`

Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/ChatChoice.java:53]: Ah. There is a `ChatRequestMessage`.

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/ChatCompletionsWithYourData.java [hunk 5]
```diff
@@ -3,14 +3,7 @@
...
 
-import com.azure.ai.openai.models.AzureChatExtensionConfiguration;
-import com.azure.ai.openai.models.AzureChatExtensionType;
-import com.azure.ai.openai.models.AzureCognitiveSearchChatExtensionConfiguration;
-import com.azure.ai.openai.models.ChatChoice;
-import com.azure.ai.openai.models.ChatCompletions;
-import com.azure.ai.openai.models.ChatCompletionsOptions;
-import com.azure.ai.openai.models.ChatMessage;
-import com.azure.ai.openai.models.ChatRole;
+import com.azure.ai.openai.models.*;
```
Comment (mssfang) [inline sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/ChatCompletionsWithYourData.java:6]: expand this import

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionCallSample.java [hunk 6]
```diff
@@ -3,15 +3,7 @@
...
-import com.azure.ai.openai.models.ChatChoice;
-import com.azure.ai.openai.models.ChatCompletions;
-import com.azure.ai.openai.models.ChatCompletionsOptions;
-import com.azure.ai.openai.models.ChatMessage;
-import com.azure.ai.openai.models.ChatRole;
-import com.azure.ai.openai.models.CompletionsFinishReason;
-import com.azure.ai.openai.models.FunctionCall;
-import com.azure.ai.openai.models.FunctionCallConfig;
-import com.azure.ai.openai.models.FunctionDefinition;
+import com.azure.ai.openai.models.*;
```
Comment (mssfang) [inline sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionCallSample.java:6]: same here. Don't use *

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/StreamingChatSample.java [hunk 7]
```diff
@@ -3,10 +3,7 @@
 
 package com.azure.ai.openai;
 
-import com.azure.ai.openai.models.ChatCompletions;
-import com.azure.ai.openai.models.ChatCompletionsOptions;
-import com.azure.ai.openai.models.ChatMessage;
-import com.azure.ai.openai.models.ChatRole;
+import com.azure.ai.openai.models.*;
```
Comment (mssfang) [inline sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/StreamingChatSample.java:6]: image

Comment (jpalvarezl) [reply sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/StreamingChatSample.java:6]: Thank you, Shawn! I somehow always forget about this setting!

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/OpenAIClient.java (hunk 2 above)
Comment (jpalvarezl) [reply sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/OpenAIClient.java:1490]: This is a good reminder that we should update the release notes too!

**Review by mssfang — APPROVED** [review]

File: sdk/openai/azure-ai-openai/CHANGELOG.md [hunk 8]
```diff
@@ -4,6 +4,37 @@
...
+| Data Source | Authentication Types |
+|---|---|
+| Azure Cognitive Search | API Key, System Managed Identity, User Managed Identity |
+| Cosmos DB (MongoDB vCore) | Connection String |
+| Pinecone | API Key |
+| Elasticsearch | Key and Key ID |
+| Azure Machine Learning Index | System Managed Identity, User Managed Identity, Access Token |
+| Microsoft Search | Access Token |
+
+To use these new sources, provide an instance of the appropriate derived type of `OnYourDataChatExtensionConfiguration` for the desired data source and use a supported `OnYourDataAuthentication` instance for its `Authentication`.
```
Comment (mssfang) [inline sdk/openai/azure-ai-openai/CHANGELOG.md:28]:
```suggestion
To use these new sources, provide an instance of the appropriate derived type of `OnYourDataVectorizationSource` for the desired data source and use a supported `OnYourDataAuthenticationOptions` instance for its `Authentication`.
```

File: sdk/openai/azure-ai-openai/CHANGELOG.md [hunk 9]
```diff
@@ -4,6 +4,37 @@
...
+
+#### dall-e-3 image generation
+
+Azure OpenAI now supports image generation with the improved `dall-e-3` models and this library supports the corresponding new features for quality level selection, prompt adjustment, and style. Use `OpenAIClient.GetImageGenerations()` and note that Azure OpenAI will *only* support `dall-e-3` models moving forward.
```
Comment (mssfang) [inline sdk/openai/azure-ai-openai/CHANGELOG.md:32]:
```suggestion
Azure OpenAI now supports image generation with the improved `dall-e-3` models and this library supports the corresponding new features for quality level selection, prompt adjustment, and style. Use `getImageGenerations()` and note that Azure OpenAI will *only* support `dall-e-3` models moving forward.
```

File: sdk/openai/azure-ai-openai/CHANGELOG.md [hunk 10]
```diff
@@ -12,6 +43,24 @@ Use `getAudioTranscription` or `getAudioTranscriptionWithResponse` convenience m
 Use `getAudioTranslation` or `getAudioTranslationWithResponse` convenience methods from respective classes.
 - Removed property, ResponseError `error` from `ImageLocation` class.
```
Comment (mssfang) [inline sdk/openai/azure-ai-openai/CHANGELOG.md:44]:
```suggestion
```

File: sdk/openai/azure-ai-openai/CHANGELOG.md [hunk 11]
```diff
@@ -4,6 +4,37 @@
 
 ### Features Added
 
+#### gpt-4-vision-preview
```
Comment (mssfang) [inline sdk/openai/azure-ai-openai/CHANGELOG.md:7]: Maybe we can copy what .NET released CHANGELOG content. 

See https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/openai/Azure.AI.OpenAI/CHANGELOG.md#100-beta10-2023-12-06

Comment (jpalvarezl) [reply sdk/openai/azure-ai-openai/CHANGELOG.md:7]: Sounds good! I took it instead from the PR description but this seems better!

File: sdk/openai/azure-ai-openai/CHANGELOG.md [hunk 12]
```diff
@@ -4,18 +4,43 @@
...
+  which is an abstract, polymorphic type with concrete derivations like `ChatRequestSystemMessage` and
+  `ChatRequestUserMessage`. This requires conversion from old `ChatMessages` into the new types. While this is
+  usually a straightforward string replacement, converting a response message into a request message (e.g. when
+  propagating an assistant response to continue the conversation) will require creating a new instance of the
+  appropriate request message with the response message's data. See the examples for details.
+
+**Dall-e-3:**
+
+- Azure OpenAI now uses `dall-e-3` model deployments for its image generation API and such a valid deployment must
+  be provided into the options for the `GetImageGenerations` method to receive results.
```
Comment (mssfang) [inline sdk/openai/azure-ai-openai/CHANGELOG.md:36]:
```suggestion
  be provided into the options for the `getImageGenerations()` method to receive results.
```

Note: .NET method name starts with an uppercase letter but Java is lowercase.

File: sdk/openai/azure-ai-openai/CHANGELOG.md [hunk 13]
```diff
@@ -4,18 +4,43 @@
...
 Use `getAudioTranscription` or `getAudioTranscriptionWithResponse` convenience methods from respective classes.
 - Removed methods `getAudioTranslationAsResponseObject` and `getAudioTranslationAsResponseObjectWithResponse` from `OpenAIClient` and `OpenAIAsyncClient` classes.
 Use `getAudioTranslation` or `getAudioTranslationWithResponse` convenience methods from respective classes.
-- Removed property, ResponseError `error` from `ImageLocation` class.
+
+**`ChatMessage` changes:**
+
+- The singular `ChatMessage` type has been replaced by `ChatRequestMessage` and `ChatResponseMessage`, the former of
+  which is an abstract, polymorphic type with concrete derivations like `ChatRequestSystemMessage` and
+  `ChatRequestUserMessage`. This requires conversion from old `ChatMessages` into the new types. While this is
```
Comment (mssfang) [inline sdk/openai/azure-ai-openai/CHANGELOG.md:28]:
```suggestion
  `ChatRequestUserMessage`. This requires conversion from old `ChatMessage` into the new types. While this is
```

**Pull request merged**